test/@types/vitest.d.ts

```diff
@@ -23,175 +27,212 @@ import type { TurnMove } from "#types/turn-move";
-
-    /**
-     * Check whether a {@linkcode Pokemon} is currently fainted (as determined by {@linkcode Pokemon.isFainted}).
-     * @remarks
-     * When checking whether an enemy wild Pokemon is fainted, one must store a reference to it in a variable _before_ the fainting effect occurs.
-     * Otherwise, the Pokemon will be removed from the field and garbage collected.
-     */
-    toHaveFainted(): void;
-
-    /**
-     * Check whether a {@linkcode Pokemon} is at full HP.
-     */
-    toHaveFullHp(): void;
-    /**
-     * Check whether a {@linkcode Pokemon} has consumed the given amount of PP for one of its moves.
-     * @param moveId - The {@linkcode MoveId} corresponding to the {@linkcode PokemonMove} that should have consumed PP
-     * @param ppUsed - The numerical amount of PP that should have been consumed,
-     * or `all` to indicate the move should be _out_ of PP
-     * @remarks
-     * If the Pokemon's moveset has been set via {@linkcode Overrides.MOVESET_OVERRIDE}/{@linkcode Overrides.ENEMY_MOVESET_OVERRIDE}
-     * or does not contain exactly one copy of `moveId`, this will fail the test.
-     */
-    toHaveUsedPP(moveId: MoveId, ppUsed: number | "all"): void;
-
-    // #endregion Pokemon Matchers
-  }
+  interface Assertion<T>
+    extends GenericMatchers<T>,
+      RestrictMatcher<GameManagerMatchers, T, GameManager>,
+      RestrictMatcher<ArenaMatchers, T, GameManager>,
+      RestrictMatcher<PokemonMatchers, T, Pokemon> {}
+}
+
+/**
+ * Utility type to restrict matchers' properties based on the type of `T`.
+ * If it does not extend `R`, all methods inside `M` will have their types resolved to `never`.
+ * @typeParam M - The type of the matchers object to restrict
+ * @typeParam T - The type parameter of the assertion
+ * @typeParam R - The type to restrict T based off of
+ * @privateRemarks
+ * We cannot remove incompatible methods outright  as Typescript requires that
+ * interfaces extend solely off of types with statically known members.
+ */
+type RestrictMatcher<M extends object, T, R> = {
+  [k in keyof M]: T extends R ? M[k] : never;
+};
+// #endregion Boilerplate/Helpers
+
+// #region Generic Matchers
+interface GenericMatchers<T> {
+  /**
+   * Check whether an array contains EXACTLY the given items (in any order).
+   *
+   * Different from {@linkcode expect.arrayContaining} as the latter only checks for subset equality
+   * (as opposed to full equality).
+   *
+   * @param expected - The expected contents of the array, in any order
+   * @see {@linkcode expect.arrayContaining}
+   */
+  toEqualUnsorted: T extends (infer U)[] ? (expected: U[]) => void : never;
+
+  /**
+   * Check whether a {@linkcode Map} contains the given key, disregarding its value.
+   * @param expectedKey - The key whose inclusion is being checked
+   * @privateRemarks
+   * While this functionality _could_ be simulated by writing
+   * `expect(x.get(y)).toBeDefined()` or
+   * `expect(x).toContain[y, expect.anything()]`,
```

Comment:
Not sure which of these you meant:
```suggestion
   * `expect(x).toContain(y, expect.anything())`,
```
```suggestion
   * `expect(x).toContain([y, expect.anything()])`,
```